src/Microsoft.Health.Fhir.SqlServer/Features/Search/Expressions/Visitors/UntypedReferenceRewriter.cs

```diff
@@ -0,0 +1,112 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Health.Fhir.Core.Features.Search.Expressions;
+using Microsoft.Health.Fhir.ValueSets;
+
+namespace Microsoft.Health.Fhir.SqlServer.Features.Search.Expressions.Visitors
+{
+    /// <summary>
+    /// For expressions over a reference search parameter where the type is not specified
+    /// (e.g. "abc" instead of "Patient/123") AND the search parameter is constrained to be
+    /// of only one type, rewrites the expression to specify the single target type.
+    /// </summary>
+    internal class UntypedReferenceRewriter : ExpressionRewriterWithInitialContext<object>
```

Comment:
>UntypedReferenceRewriter [](start = 19, length = 24)

nit: I would consider adding `SearchParameter` to name since it's operates only on search parameters.
Or maybe it's too mouthful...

Reply:
All parsed expressions are either a single SearchParameterExpression or ANDed SearchParameterExpressions, so I'm not sure how helpful the addind that to the name would be. (A later rewriter creates a SqlRootExpression, but ultimately, any expression over any field is always going to be under a SearchParameterExpression).